src/coreclr/gc/gc.cpp

```diff
@@ -17976,6 +17944,63 @@ size_t gc_heap::current_generation_size (int gen_number)
     return gen_size;
 }
 
+#ifdef USE_REGIONS
+// We may need a new empty region while doing a GC so try to get one now, if we don't have any
+// reserve in the free region list. 
+bool gc_heap::try_get_new_free_region()
+{
+    heap_segment* region = 0;
+    if (free_regions)
+    {
+        dprintf (REGIONS_LOG, ("h%d has %d free regions %Ix", heap_number, num_free_regions,
+            heap_segment_mem (free_regions)));
+        return true;
+    }
+    else
+    {
+        region = allocate_new_region (__this, 0, false);
+        if (region)
+        {
+            init_table_for_region (0, region);
+            return_free_region (region);
+            dprintf (REGIONS_LOG, ("h%d got a new empty region %Ix", heap_number, region));
+        }
+    }
+
+    return (region != 0);
+}
+
+void gc_heap::init_table_for_region (int gen_number, heap_segment* region)
+{
+#ifdef BACKGROUND_GC
+        if (is_bgc_in_progress())
+        {
+            dprintf (GC_TABLE_LOG, ("new seg %Ix, mark_array is %Ix", 
+                heap_segment_mem (region), mark_array));
+            if (!commit_mark_array_new_seg (__this, region))
+            {
+                dprintf (GC_TABLE_LOG, ("failed to commit mark array for the new region %Ix-%Ix",
+                    get_region_start (region), heap_segment_reserved (region)));
+
+                // We don't have memory to commit the mark array so we cannot use the new region.
+                global_region_allocator.delete_region (get_region_start (region));
+                region = 0;
+            }
+        }
+#endif //BACKGROUND_GC
+
+    if (gen_number <= max_generation)
+    {
+        size_t first_brick = brick_of (heap_segment_mem (region));
```

Comment:
Would this AV if commit_mark_array_new_seg failed above?

Reply:
it wouldn't - right now the GC bookkeeping datastructure is completely committed for the region's range except for the mark array part.

Reply:
but above, you set region to 0, so as heap_segment_mem dereferences it, you would AV before you even get to the brick_of part.

Reply:
oh I misunderstood your intention for the question...absolutely, I shouldn't call this when the region is already set to 0